### ISSUE TITLE
Feats tab discussion

### ISSUE THREAD

**Issue by Heliomance** [Heliomance/HeroForge-Anew#68]
The Feats tab is a problem. There's somewhere in the vicinity of 7000 checkboxes on it, and that makes it really slow and sluggish. I'm open to the idea of redesigning it somehow, and I'm opening this issue to discuss what I might do with it.

Currently, my only idea is to replace it with a series of dropdown boxes for each level where you gain a feat, allowing you to choose a feat for that level. That has a lot of issues, though. Different characters have different numbers of feats, and designing a layout that could work would be hard. It would probably need 40-50 drop down boxes, each linked to a separate data list tracking the valid feats to take at that level, and able to cope with limited bonus feat lists - and that's without Epic support. It would also be very inflexible, unable to cope with a character getting more feats in a level than it's set up to handle, so in all I think it's probably not a great idea. Does anyone have any alternatives?

**Comment by TwistedFaith84** [comment]
While you could just get rid of the bonus feats column altogether and just have the total number of feats being counted and that would get rid of exactly half of the checkboxes. and then to show bonus feats and free feats still colored properly go with something like what I pointed out here
_**https://github.com/Heliomance/HeroForge-Anew/issues/53#issuecomment-164252364**_. Though I don't know how much if any it's gonna speed up the Feats sheet but dropping @ least half of the checkboxes is a very good start. And for switching them out to dropdown boxes is just switching out one devil for another. I mean sure you can just make sections based on levels and then have enough spaces in those sections to allow for the Maximum amount of feats someone can gain per level but that can be tricky as you said.

**Comment by TwistedFaith84** [comment]
Just tried the deleting of the checkboxes in the bonus feats column (I'm not savvy enough to try the other part of my idea) anyways it seemed to help. It sped up the first loading of the feats page a little bit. Scrolling threw the Feats Sheet and the time it take to recalculate Is **ALOT** faster then it was. This was tested in Excel 2010 so I don't know how much of a difference it would make on Excel 2013 or above.

**Comment by WesleyVos** [comment]
Any word on a fix for this yet? Even deleting the bonus feat column doesn't do enough. It's still slow as can be scrolling up or down on the list.

**Comment by TwistedFaith84** [comment]
@WesleyVos What version of Excel are you using? And when you deleted the bonus feats column did you just highlight the column and delete that column or did you actually select all the checkboxes in the bonus feat column and delete those?

**Comment by WesleyVos** [comment]
@TwistedFaith84 I'm using Office 365, so it's updated to Excel 2016. And I deleted the entire column. Trying to scroll down to delete all the checkboxes individually would have taken far more time than I have available.

What's odd is that before I upgraded from 10 to 16, I had no issues with the sheet.

**Comment by TwistedFaith84** [comment]
@WesleyVos Yea Because you just deleted the Column it doesn't actually delete any of the checkboxes in that column. In order to delete a large number of **OBJECTS** if you look at the **Find & Select** you can see an option that says **Select Objects** then just Drag Select the column and you can delete Most if not All the checkboxes in that column. I know that it will take a while to do on everything above 2010 but that the best way instead of selecting all **OBJECTS** one at a time. Just remember to unselect the **Select Objects** in the **Find & Select** when your done.

**Comment by torrasque666** [comment]
Figured out a solution. Shamelessly cribbing from Sagasheet.

Remove all the objects in the main section (everything below the frozen view area). All of them. As a general rule, change the format of all cells where the old checkboxes were to a custom format that reads "Yes";"No";"No"

This means that any value greater than 0 will read as "Yes". Change the old linked cells for the checkboxes (O11:P3634) to "=IF(B11>=1,TRUE,FALSE)" and paste that across all of the cells. Bam. Any feat that can be selected multiple times can have its value limited through the Data Validation option (limited to whole numbers, from 0 to X where X is the max, if needed, like for Martial Study) and instead have ###;"No";"No" for its format.

We'll need to make a new sheet to cover the various feats that used to be handled by a dropdown. Though then it becomes a question of how many to use. 3? 5? 10?

**Comment by Heliomance** [comment]
...I've just looked at how YAPCG does feats. It is so unbelievably better. I want to shift over HeroForge to doing it like that, but that's going to be a hell of a lot of work. This puts me in a conundrum.

I was hoping that v8.0 was going to be the last save-breaking update, planning to change save files to working by XML. Changing over to YAPCG-style feats is going to be a save-breaking update no matter what I do, I'm fairly sure. So do I add it to v8.0, which will delay it even more than its current "when it happens" timeframe, or do I do v8.0 without changing the feats, accepting that that means there's going to be another save-breaking update further down the line?

**Comment by torrasque666** [comment]
I tried that. 3.5 has literally thousands of feats compared to pathfinder's hundreds. Much more processing power, a crap ton more bonus lists, etc. Add in the fact that there's potentially 3 feats a level (HD, class boni, vow of poverty/elder evil worship) with an additional 3 at first (human and flaws) and that hero forge is expected to handle up to 60 levels (unless we're gonna slash that) and you wind up with close to 200 feat slots.

**Comment by Heliomance** [comment]
You might get three feats on any given level, but I don't expect you'd ever get three feats at every level. And the beauty of the way YAPCG does it is that it assigns the feat slots dynamically. It also does the drop-downs with data validation instead of form controls, which I'm fairly sure is less resource intensive. I'm pretty certain it's going to be fairly easy to extend, as well, so once the framework is in place it's probably easy enough to have as many feat slots as you like.

The number of possible different bonus feat lists is a problem, but not an insurmountable one - especially if I can work out a way to make it auto-populate the lists from the existing bonus feat system, which I think I probably can. If not, then it's just a lot of setup work. The sheer number of feats is a potentially bigger problem, but again I'd try to adapt as much as possible of the existing infrastructure.

Basically I think it's probably doable, but it would take time. And the question is whether I spend that time before or after v8.0 is released.

**Comment by torrasque666** [comment]
My test assigned them dynamically. The number of feats was more "how many rows do I need to potentially set aside for selection." It was more just the sheer number. When I get home and off mobile I can give you a link to my old test screenshots, as well as my old test document if you'd like.

**Comment by Keldin** [comment]
I primarily play Pathfinder these days, so I use YAPCG for that.

I don't know if this helps any, but one complaint I did have with YAPCG's feats tab was there wasn't enough of them, at least for one of my characters.  The artificer (http://www.d20pfsrd.com/classes/3rd-party-classes/adamant-entertainment/artificer) gains five bonus feats and eight item creation feats across twenty levels.  There isn't enough room for all of those on YAPCG's feats tab, though it's a fairly easy fix - you can just enter them separately, as custom text or part of the custom class.  It just annoys me that that doesn't put them in the feat section with the rest of the feats.

Joel

**Comment by torrasque666** [comment]
To be fair Joel, YAPCG is only inherently designed for 1st party content. It has some custom content support but not much. Its not suprising that something that gains an additional 13 feats would exceed the limits of the program. Granted, fighter gets 11 over 20 levels but thats probably the max it was intended to work with.

**Comment by Keldin** [comment]
Oh, I know.  I wasn't complaining, not really, more thinking that it might be considered an example of the need for additional "padding" in the available area.  After all, the first-party content that Adamant Entertainment's class is based on - the Artificer from Eberron - ALSO has eight item creation feats and five bonus feats.

Joel

**Comment by torrasque666** [comment]
#50 was the one where i had posted [this](https://github.com/Heliomance/HeroForge-Anew/issues/50#issuecomment-134881389), which was an example of trying to make it more like YAPCG's feat section

and here's the dropbox link (warning, its frikkin massive) https://www.dropbox.com/s/660ktndb6doxkww/HeroForge%20Anew%203.5%20v7.4.0.1%20feats%20test.xls?dl=0

**Comment by TwistedFaith84** [comment]
It's been a while since I used or looked at YAPCG but what little I do remember about it's feat section was yea that it was smaller in both stature and substance, and like @Keldin said you really didn't  have enough room for the amounts of feats you can get with certain classes or combos (ie a Gestalted Human Fighter/Rogue (Martial ACF) right there is a min of 30 Feats (22 are bonus) at level 20 and thats b4 any flaws. I just checked that YAPCG is currently only set up to have a max of 38 feats (10 are bonus) at level 20.

**Comment by TwistedFaith84** [comment]
Just had an Idea why not just keep it listed the way it is but get rid of all the checkboxes for feats all together and switch them out with someone typing a number especially for feats that can be taken more then once. The numbers could range from 1-10 (don't know why someone would want to take the same feat more the 10x), feats that can be taken only once would be just 1, and feats that can be taken more then once would be 1-10 and could be handled either on a separate sheet or just stick with dropdowns (just add a few more dropdowns for the ones that need it). No granted this would alleviate most of the form controlled items and just switch them out with either more formulas and/or VBA.

PS cells with no numbers would mean you didn't select the feat and as far as the dropdowns go for the feats that can be chosen multiple times have the first item listed as "Select" and if something is selected and there is a the appropriate number in the cell thats the feat you selected and and the appropriate item for that feat so many times.

**Comment by Heliomance** [comment]
Because that would be significantly less user friendly. 

I'm definitely leaning towards the YAPCG solution. My poke  through yesterday didn't show an excessive quantity of array formulae, and as I say it seems quite extensible. I'd have to look into it more closely to say for sure if it's feasible, but if it is, it shouldn't be much harder to add 50 slots than it is to add 5.

**Comment by TwistedFaith84** [comment]
I don't see how that would be less user friendly as that's really how it is now; just with form controlled checkboxes instead (1=checked and blank=unchecked). Although for the feats that can be taken multiply times the numbers would just mean how many times your taking that feat. Sure it might mean a little more coding but its gonna be a lot more coding the other the way YAPCG has it cause you would definitely be starting from scratch to where with my suggestion most of the work is already done. Anyways yea if you still want to do it the way YAPCG has it your right after you look into it; adding 50+ slots shouldn't be much harder then adding 5 (you would need to definitely need to add more then 50 slots if you still wanted to keep HF able to go upto lvl 60)

*_EDIT:: *_Just realized the my suggestion was basically the same as @torrasque666 did here

> (https://github.com/Heliomance/HeroForge-Anew/issues/68#issuecomment-174043159)

**Comment by TwistedFaith84** [comment]
I will say that the reason I like how the Feats listing is now is because without having to select the feat to have it add to what you selected for your feats you get a short description of what the feat actually does where as the way that YAPCG has it you have to select the feat to see what the feat does. Which even for experienced player slows down your character creation by having to go to the dropdown and select the feat so you can see what it does (if it's a feat your not accustomed to) then having to go back to the dropdown and select another feat to see what it does so on and so forth until you find a feat you wanna use so that one feat and then having to do it all again for the rest of your feats where as the way it is currently listing all you have to do is highlight the feat name and you can see what it does then simply use the down arrow and you can find your feats 3x-5x faster cause all you have to do after you find the feat you want to use is click the box (right now to check it) and type in a number for how many times you want that feats (using mine and kinda @torrasque666  suggested methods) and for your next feat your still in the same place you left off at so just keep using the down arrow to find your next feat or scroll back up to a feat you saw b4 and quickly select that and continue until all your feats are selected.

**Comment by ThAlEdison** [comment]
You could use data validation to restrict feats that can't be taken more
than once.

~Erick A. Cobb

**Comment by TwistedFaith84** [comment]
@ThAlEdison yea that can be used to do that as well as restrict the maximum number of times feats can be taken multiple times (something like 10-15)
